contracts/staking/stakeManager/StakeManager.sol

```diff
@@ -631,38 +683,288 @@ contract StakeManager is IStakeManager, StakeManagerStorage, Initializable {
+        uint256 reward = blockInterval.mul(CHECKPOINT_REWARD).div(checkPointBlockInterval);
+        reward = reward.mul(signedStakePower).div(currentTotalStake);
+        reward = Math.min(CHECKPOINT_REWARD, reward);
+
+        uint256 _proposerBonus = reward.mul(proposerBonus).div(MAX_PROPOSER_BONUS);
+        uint256 proposerId = signerToValidator[proposer];
+
+        Validator storage _proposer = validators[proposerId];
+        uint256 delegatedAmount = _proposer.delegatedAmount;
+        if (delegatedAmount > 0) {
+            _increaseValidatorRewardWithDelegation(proposerId, _proposer.amount, delegatedAmount, _proposerBonus);
+        } else {
+            _proposer.reward = _proposer.reward.add(_proposerBonus);
+        }
+
+        // update stateMerkleTree root for accounts balance on heimdall chain
+        accountStateRoot = stateRoot;
+
+        uint256 newRewardPerStake = rewardPerStake.add(
+            reward.sub(_proposerBonus).mul(REWARD_PRECISION).div(signedStakePower)
+        );
+
+        // evaluate rewards for validator who did't sign and set latest reward per stake to new value to avoid them from getting new rewards.
+        _updateValidatorsRewards(unsignedValidators, totalUnsignedValidators, newRewardPerStake);
+
+        // distribute rewards between signed validators
+        rewardPerStake = newRewardPerStake;
+
+        // evaluate rewards for unstaked validators to avoid getting new rewards until they claim their stake
+        _updateValidatorsRewards(deactivatedValidators, totalDeactivatedValidators, newRewardPerStake);
+
+        _finalizeCommit();
+        return reward;
+    }
+
+    function _updateValidatorsRewards(
+        uint256[] memory unsignedValidators,
+        uint256 totalUnsignedValidators,
+        uint256 newRewardPerStake
+    ) private {
+        uint256 currentRewardPerStake = rewardPerStake;
+        for (uint256 i = 0; i < totalUnsignedValidators; ++i) {
+            _updateRewardsAndCommit(unsignedValidators[i], currentRewardPerStake, newRewardPerStake);
+        }
+    }
+
+    function _updateRewardsAndCommit(
+        uint256 validatorId,
+        uint256 currentRewardPerStake,
+        uint256 newRewardPerStake
+    ) private {
+        uint256 validatorsStake = validators[validatorId].amount;
+        uint256 delegatedAmount = validators[validatorId].delegatedAmount;
+        if (delegatedAmount > 0) {
+            uint256 combinedStakePower = validatorsStake.add(delegatedAmount);
+            _increaseValidatorRewardWithDelegation(
+                validatorId,
+                validatorsStake,
+                delegatedAmount,
+                _getEligibleValidatorReward(validatorId, combinedStakePower, currentRewardPerStake)
+            );
+        } else {
+            _increaseValidatorReward(
+                validatorId,
+                _getEligibleValidatorReward(validatorId, validatorsStake, currentRewardPerStake)
+            );
+        }
```

Comment:
Can't we just skip this code block if `currentRewardPerStake <= rewardsPerStake`? And skip directly to `validators[validatorId].initialRewardPerStake = newRewardPerStake`?

Reply:
I can skip it yes, will do.